File: crates/chidori-js/benchmarks/workloads/mixed_helpers.js
// Mixed "glue code": small helper functions over objects and strings — the
// shape of real agent/tool plumbing that the typed kernel tiers deliberately
// do NOT cover (non-numeric bodies, string building, property traffic across
// calls, for-in). This is the register-bytecode tier's home turf: with loop
// kernels declining every function here, the interpreter itself carries the
// whole workload.
const N = 60_000;

function normalize(rec) {
  const out = { id: rec.id | 0, label: "", score: 0 };
  for (const k in rec) {
    if (k === "id") continue;
    const v = rec[k];
    if (typeof v === "number") out.score = out.score + v;
    else if (typeof v === "string") out.label = out.label ? out.label + ":" + v : v;
  }
  return out;
}

function render(rec) {
  return "[" + rec.id + "] " + (rec.label || "?") + " => " + rec.score;
}

function classify(score) {
  return score > 40 ? "hi" : score > 15 ? "mid" : "lo";
}

const buckets = { hi: 0, mid: 0, lo: 0 };
let text = "";
let checksum = 0;
for (let i = 0; i < N; i++) {
  const rec = { id: i, name: "item" + (i % 7), a: i % 13, b: (i * 3) % 29, kind: i % 2 ? "x" : "y" };
  const norm = normalize(rec);
  const line = render(norm);
  buckets[classify(norm.score)]++;
  checksum = (checksum + line.length + norm.score) % 1000000007;
  if ((i & 1023) === 0) text += line + "\n";
}
console.log("RESULT=" + checksum + "|" + buckets.hi + "," + buckets.mid + "," + buckets.lo + "|" + text.length);

File: crates/chidori-js/benchmarks/workloads/mixed_helpers.py
# Python twin of mixed_helpers.js — small helper functions over dicts and
# strings, the shape of real agent/tool plumbing: key iteration, per-value
# type dispatch, string building, and dict property traffic across calls.
# All strings are ASCII, so Python len() agrees with JS .length.
N = 60_000


def normalize(rec):
    out = {"id": rec["id"], "label": "", "score": 0}
    for k in rec:
        if k == "id":
            continue
        v = rec[k]
        if isinstance(v, int):
            out["score"] = out["score"] + v
        elif isinstance(v, str):
            out["label"] = out["label"] + ":" + v if out["label"] else v
    return out


def render(rec):
    return "[" + str(rec["id"]) + "] " + (rec["label"] or "?") + " => " + str(rec["score"])


def classify(score):
    return "hi" if score > 40 else "mid" if score > 15 else "lo"


buckets = {"hi": 0, "mid": 0, "lo": 0}
text = ""
checksum = 0
for i in range(N):
    rec = {
        "id": i,
        "name": "item" + str(i % 7),
        "a": i % 13,
        "b": (i * 3) % 29,
        "kind": "x" if i % 2 else "y",
    }
    norm = normalize(rec)
    line = render(norm)
    buckets[classify(norm["score"])] += 1
    checksum = (checksum + len(line) + norm["score"]) % 1000000007
    if (i & 1023) == 0:
        text += line + "\n"
print(
    "RESULT="
    + str(checksum)
    + "|"
    + str(buckets["hi"])
    + ","
    + str(buckets["mid"])
    + ","
    + str(buckets["lo"])
    + "|"
    + str(len(text))
)
